crates/kitsune_p2p/fetch/src/queue.rs

```diff
@@ -0,0 +1,328 @@
+    /// If the FetchKey exists, add the new source and merge the context in, without
+    /// changing the position in the queue.
+    pub fn push(&mut self, request: FetchRequest, space: KSpace, agent: KAgent) {
+        self.state
+            .share_mut(|s, _| Ok(s.push(&*self.config, request, space, agent)))
+            .expect("no error");
+    }
+
+    /// When an item has been successfully fetched, we can remove it from the queue.
+    pub fn remove(&mut self, key: &FetchKey) {
+        self.state
+            .share_mut(|s, _| Ok(s.remove(key)))
+            .expect("no error");
+    }
+}
+
+impl State {
+    /// Add an item to the queue.
+    /// If the FetchKey does not already exist, add it to the end of the queue.
+    /// If the FetchKey exists, add the new source and merge the context in, without
+    /// changing the position in the queue.
+    pub fn push(
+        &mut self,
+        config: &dyn FetchQueueConfig,
+        request: FetchRequest,
+        space: KSpace,
+        agent: KAgent,
+    ) {
+        let FetchRequest {
+            key,
+            author,
+            options,
+            context,
+        } = request;
+
+        match self.queue.entry(key) {
+            Entry::Vacant(e) => {
+                let sources = if let Some(author) = author {
+                    Sources(vec![Source::new(agent), Source::new(author)])
+                } else {
+                    Sources(vec![Source::new(agent)])
+                };
+                let item = FetchQueueItem {
+                    sources,
+                    space,
+                    options,
+                    context,
+                };
+                e.insert(item);
+            }
+            Entry::Occupied(mut e) => {
+                let v = e.get_mut();
+                v.sources.0.insert(0, Source::new(agent));
+                v.options = options;
+                v.context = match (v.context.take(), context) {
+                    (Some(a), Some(b)) => Some(config.merge_contexts(*a, *b).into()),
+                    (a, b) => a.and(b),
+                }
+            }
+        }
+    }
+
+    /// Access queue items through mutable iteration. Items accessed will be moved
+    /// to the end of the queue.
+    ///
+    /// Only items whose `last_fetch` is more than `interval` ago will be returned.
+    pub fn iter_mut(&mut self, interval: Duration) -> StateIter {
+        StateIter {
+            state: self,
+            interval,
+        }
+    }
+
+    /// When an item has been successfully fetched, we can remove it from the queue.
+    pub fn remove(&mut self, key: &FetchKey) {
+        self.queue.remove(key);
+    }
+}
+
+impl<'a> Iterator for StateIter<'a> {
+    type Item = (FetchKey, KSpace, KAgent);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let keys: Vec<_> = self
+            .state
+            .queue
+            .keys()
+            .take(NUM_ITEMS_PER_POLL)
```

Comment:
What happens in the edge case where we have more than this number of items at the front of the queue that all return `None` for the `sources.next`?

Reply:
@neonphog, then the iterator will return `None` for that round, but all of those keys will move to the back of the map, so on the next iteration, fresh items will be inspected.

Maybe this could be done better though. This implementation I think makes sure we never get stuck, since items are constantly rotating through the queue, but perhaps we could be a bit more methodical.